Feature/#34 구독 페이지 UI 구현

soo96 opened the pull request:
## 요약
- 구독 페이지 UI 구현

Close: #34

## 변경 사항
### 1️⃣ 구독 요청
- [x] 구독할 URL 을 입력받을 수 있는 `input` 
- [x] 구독 요청을 보낼 수 있는 `[확인]` 버튼
- [x] 구독 API 요청 기능
- [x] 성공 모달
  - [x] "구독 채널이 추가되었습니다."
- [x] 실패 모달
  - [x] "이미 구독중인 채널입니다."
  - [x] "더 이상 채널을 구독할 수 없습니다."
  - [x] "구독 채널 추가에 실패했습니다. URL을 확인해주세요."

### 2️⃣ 구독목록 조회
- [x] 사용자가 구독한 채널목록 정보 보여주기
    - [x] 채널명
    - [x] 플랫폼명
    - [x] 구독 시작 일시

### 3️⃣ 구독 취소
- [x] 구독 취소 API 요청 기능
- [x] 구독 취소 요청을 보낼 수 있는 `[-]` 버튼 
  - [x]  버튼 클릭시 컨펌 모달 표출
    - [x] 채널 구독을 해지하시겠습니까?
- [x] 성공 모달
  - [x] "채널 구독이 해지되었습니다."
- [x] 실패 모달
  - [x] "구독채널 해지 도중 문제가 발생했습니다."

### ☑️ 추가: 공통 `ConfirmModal`, `AlertModal`  구현

## 체크리스트

- [x] 이해하기 어려운 코드에는 주석을 추가했습니다  
- [x] 관련 문서를 수정했습니다  
- [x] 새로운 경고가 발생하지 않습니다  
- [x] 수정한 기능 또는 추가한 기능에 대해 테스트를 작성했습니다  
- [x] 의존성 있는 변경 사항이 병합 및 배포되었습니다

## 관련 이슈

- #34 
- #23 

## 리뷰 반영 사항

<!-- PR리뷰 반형 사항 -->

<details>
<summary><strong>선택 항목 펼치기</strong></summary>

## 스크린샷

<!-- 시각적인 변경 사항이 있다면 스크린샷이나 GIF를 포함해 주세요. 리뷰어가 쉽게 이해할 수 있습니다. -->

## 테스트 방법

<!-- PR에서 수행한 변경 사항을 어떻게 테스트할 수 있는지 설명해 주세요. 리뷰어가 직접 확인할 수 있도록 돕습니다. -->

## 리뷰어 참고사항

<!-- 리뷰어에게 특별히 알려야 할 사항이나 고려할 점이 있다면 여기에 작성해 주세요. -->

</details>

soo96 committed:
feat: InputWithButton 공통 컴포넌트 추가

soo96 committed:
feat: 구독 페이지 정적 UI 컴포넌트 구현 (Subscription, SubscriptionItem)

soo96 committed:
chore: Header 컴포넌트 구독 타이틀 변경

soo96 committed:
feat: Input에서 Enter 키 입력 시 제출 기능 추가

soo96 committed:
feat: 공통 Confirm, Alert 모달 컴포넌트 구현

soo96 committed:
feat: 구독 요청 및 구독목록 조회 기능 구현

soo96 committed:
refactor: 구독 요청/조회 로직 useSubscribeMutation, useSubscriptionQuery 커스텀 훅으로 분리

soo96 committed:
feat: 구독 취소 기능 구현

soo96 committed:
refactor: 구독 취소 로직 useUnsubscribeMutation 커스텀 훅으로 분리

soo96 committed:
chore: 구독페이지 안내문구 추가 및 스크롤 및 레이아웃 관련 스타일 수정

soo96 committed:
test: 구독페이지 추가로 인한 App.spec.tsx 테스트 코드 수정

soo96 committed:
design: 구독 해지 버튼 아이콘 스타일 변경

leesm0218 commented on src/constants/styles.ts:
```diff
@@ -142,3 +142,45 @@ export const SIGNUP_PAGE_STYLE = {
+  input:
+    'w-full rounded-xl text-lg px-4 py-2 border border-gray-300 focus:outline-none focus:ring-2 focus:ring-indigo-500',
+  button:
+    'bg-purple-600 text-white font-bold rounded-md w-24 h-11 hover:bg-purple-700 transition"',
```
```suggestion
    'bg-purple-600 text-white font-bold rounded-md w-24 h-11 hover:bg-purple-700 transition',
```

soo96 replied:
감사합니다.

smileboy0014 commented on src/components/subscription/SubscriptionItem.tsx:
```diff
@@ -0,0 +1,82 @@
+
+  const confirmMessage = `[${channel.channelName}] 채널 구독을 해지하시겠습니까?`;
+
+  const { mutate: mutateUnsubscribe } = useUnsubscribeMutation(
+    () => setAlertMessage('채널 구독이 해지되었습니다.'),
+    () => setAlertMessage('구독채널 해지 도중 문제가 발생했습니다.')
+  );
```
custom hook으로 빼서 관심사의 분리를 한 건 너무 좋은 것 같습니다!
다만, 처음 이 코드를 보는 경우, 콜백으로 어떤 걸 처리하는지 바로 보이지 않으니, 이렇게 해보는건 어떨까요??
```suggestion
  const { mutate: mutateUnsubscribe } = useUnsubscribeMutation(
    onSuccess: () => setAlertMessage('채널 구독이 해지되었습니다.'),
    onError: () => setAlertMessage('구독채널 해지 도중 문제가 발생했습니다.')
  );
```
가독성도 한번 생각해보면 좋을 것 같습니다!

soo96 replied:
리뷰 감사합니다!
다시 보니 함수가 선언된 부분을 확인하지 않고는 어떤 역할을 하는 함수를 인자로 넘기는지 바로 확인이 어렵긴 하겠네요.
```suggestion
  const { mutate: mutateUnsubscribe } = useUnsubscribeMutation({
    onSuccess: () => setAlertMessage('채널 구독이 해지되었습니다.'),
    onError: () => setAlertMessage('구독채널 해지 도중 문제가 발생했습니다.'),
  });
```
이렇게 객체로 전달하는 방식 말쓸하시는 걸까요?

smileboy0014 replied:
네!! 어떤걸 콜백으로 넘기는지 알면 더 좋을 것 같습니다 :)

smileboy0014 commented on src/components/subscription/SubscriptionItem.tsx:
```diff
@@ -0,0 +1,82 @@
+            <span className={SUBSCRIPTION_ITEM_STYLE.dateText}>
+              {subscribedAt}
+            </span>
+            <button
+              onClick={() => handleClickButton()}
+              className={SUBSCRIPTION_ITEM_STYLE.deleteButton}
+            >
```
```suggestion
            <button
              onClick={() => handleClick()}
              className={SUBSCRIPTION_ITEM_STYLE.deleteButton}
            >
```
물론 `handleClickButton` 함수가 더 명확하지만, 클릭 버튼이 이 컴포넌트에 여러개 있는게 아니라면
```javascript
on[EventName] <=> handle[EventName]
```
를 지켜주는게 더 좋지 않을까요?

soo96 replied:
```suggestion
            <button
              onClick={() => setShowConfirmModal(true)}
              className={SUBSCRIPTION_ITEM_STYLE.deleteButton}
            >
```
@smileboy0014 
함수 로직이 setShowConfirmModal(true)뿐이라서 별도 handle함수로 분리할 필요가 없을 것 같은데 어떠실까요?

smileboy0014 commented on src/pages/Home/Subscription.tsx:
```diff
@@ -0,0 +1,80 @@
+    }
+  );
+
+  const { data, isLoading } = useSubscriptionsQuery();
```
```suggestion
  const { data, isError, isLoading } = useSubscriptionsQuery();
  
  if (isError) {
      return <div> 데이터를 가져오지 못 했습니다 </div>
  }
```
만약 가져오는 쿼리가 문제가 생겼을 때, fallback 처리를 해야하지 않을까요??

soo96 replied:
리뷰 감사합니다! 😊
fallback 처리 추가하면 좋을 것 같습니다.

@leesm0218 @smileboy0014 
추가로 useSuspenseQuery()를 사용하여 로딩과 에러를 로딩/에러 상태를 UI 컴포넌트 밖에서 공통적으로 처리할 수 있다고 합니다.
컴포넌트 내부 코드가 깔끔해지고, 관심사의 분리 측면의 장점도 생길 것 같네요.
```ts
<ErrorBoundary FallbackComponent={<Fallback />}> // 에러상태 분리
  <Suspense fallback={<Loading />}> // 로딩상태 분리
    <MyComponent />
  </Suspense>
</ErrorBoundary>
```
아래 관련 링크 달아두겠습니다!

### Ref
- https://ko.react.dev/reference/react/Suspense
- https://tanstack.com/query/latest/docs/framework/react/reference/useSuspenseQuery
- https://lasbe.tistory.com/204
- #23

smileboy0014 replied:
아 그렇게 많이 처리를 하는군요! 저걸 도입하면 공통으로 fallback 처리를 할 수 있어서 더 코드가 정말 깔끔해질 것 같습니다! 그럼 #23 이슈에 적용해서 테스크 진행하면 좋을 것 같습니다 😀

smileboy0014 commented on src/components/subscription/SubscriptionItem.tsx:
```diff
@@ -0,0 +1,82 @@
+        <ConfirmModal
+          message={confirmMessage}
+          onConfirm={handleConfirm}
+          onCancel={() => setShowConfirmModal(false)}
```
```suggestion
          onCancel={handleCloseModal}
```
```javascript
const handleCloseModal = () => setAlertMessage('');
<AlertModal message={alertMessage} onClose={handleCloseModal} />
```
밑에 [src/pages/Home/Subscription.tsx] 에서는 이렇게 한 줄이더라도 handler 함수로 빼셨더라구요!

그런데 또 여기서는 handler 함수로 안 빼고 계신 것 같아서요!

통일성이 있어야 하지 않을까요???

soo96 replied:
```ts
          onCancel={() => setShowConfirmModal(false)}
```
함수 로직이 `setShowConfirmModal(false)`뿐이라서 별도 handle함수로 분리할 필요가 없을 것 같습니다.
[src/pages/Home/Subscription.tsx]에서도 동일하게 적용하도록 하겠습니다.🙂

smileboy0014 commented on src/components/common/InputWithButton.tsx:
```diff
@@ -0,0 +1,43 @@
+        type="text"
+        value={inputText}
+        placeholder={placeholder}
+        onChange={(e) => onChange(e.target.value)}
```
```suggestion
        onChange={onChange}
```
지금 InputWithButton은 common에 위치하고 있습니다. 즉, 공통 컴포넌트로 만들어 두셨다는건데, 그럼 정말 열어두게 설계를 해야한다고 저는 생각합니다.
물론 대부분은 `e.target.value`를 쓰겠지만, 누구는 `e.target.innerText` 값을 다루고 싶을 수도 있습니다.
그러니 어느 특정 페이지에 속하는 컴포넌트가 아니고 이렇게 공통으로 열어두는 컴포넌트라면, 이 컴포넌트를 사용하는 사용자에게 조금 열어두는 방향이 더 좋지 않을까요??

soo96 replied:
리뷰 감사드립니다. 🙏🏻
경태님 말씀대로 event.target의 속성을 고정해두면 재사용성과 확장성이 떨어질 수 있겠네요.

leesm0218 reviewed (approved):
수고하셨습니다.

soo96 committed:
feat: 로그인 성공시 인증 요청을 위한 Axios 기본 헤더 설정

soo96 committed:
feat: 구독목록 조회 실패 시 fallback UI 추가

soo96 committed:
refactor: PR 리뷰 반영

- InputWithButton 컴포넌트 onChange prop 형식 변경
- Subscription, SubscriptionItem 코드 스타일 수정
- useSubscribeMutation, useUnsubscribeMutation 커스텀 훅 매개변수 객체로 한번에 받도록 수정

Refs: #34

soo96 committed:
chore: style 문자열 오타 수정

Co-authored-by: leesm0218 <leesm0218@naver.com>

smileboy0014 reviewed (approved):
코드리뷰한 거 다 반영해주셔서 감사합니다!! 고생하셨습니다!

soo96 commented:
고생하셨습니다! 감사합니다!!🙏🏻
